File: src/main/java/alfheim/api/event/LivingPotionEvent.java
package alfheim.api.event;

import cpw.mods.fml.common.eventhandler.Cancelable;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.potion.PotionEffect;
import net.minecraftforge.event.entity.living.LivingEvent;

public abstract class LivingPotionEvent extends LivingEvent {

	public final PotionEffect effect;
	
	public LivingPotionEvent(EntityLivingBase entity, PotionEffect pe) {
		super(entity);
		effect = pe;
	}

	public abstract static class Add extends LivingPotionEvent {
		public Add(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }

		@Cancelable
		@Deprecated
		public static class Pre extends Add {
			public Pre(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }
		}
		
		public static class Post extends Add {
			public Post(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }
		}
	}

	public abstract static class Change extends LivingPotionEvent {
		public Change(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }

		@Cancelable
		@Deprecated
		public static class Pre extends Change {
			public Pre(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }
		}
		
		public static class Post extends Change {
			public Post(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }
		}
	}

	public abstract static class Remove extends LivingPotionEvent {
		public Remove(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }

		@Cancelable
		@Deprecated
		public static class Pre extends Remove {
			public Pre(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }
		}
		
		public static class Post extends Remove {
			public Post(EntityLivingBase entity, PotionEffect pe) { super(entity, pe); }
		}
	}
}

File: src/api/java/alexsocol/patcher/event/LivingPotionEvent.kt
package alexsocol.patcher.event

import cpw.mods.fml.common.eventhandler.Cancelable
import net.minecraft.entity.EntityLivingBase
import net.minecraft.potion.PotionEffect
import net.minecraftforge.event.entity.living.LivingEvent

abstract class LivingPotionEvent(entity: EntityLivingBase, val effect: PotionEffect): LivingEvent(entity) {
	
	abstract class Add(entity: EntityLivingBase, pe: PotionEffect): LivingPotionEvent(entity, pe) {
		
		@Cancelable
		@Deprecated("Unimplemented", level = DeprecationLevel.ERROR)
		class Pre(entity: EntityLivingBase, pe: PotionEffect): Add(entity, pe)
		
		class Post(entity: EntityLivingBase, pe: PotionEffect): Add(entity, pe)
	}
	
	abstract class Change(entity: EntityLivingBase, pe: PotionEffect,
	                      /** Remove and add potion modifiers  */
	                      val update: Boolean): LivingPotionEvent(entity, pe) {
		
		@Cancelable
		@Deprecated("Unimplemented", level = DeprecationLevel.ERROR)
		class Pre(entity: EntityLivingBase, pe: PotionEffect, update: Boolean): Change(entity, pe, update)
		
		class Post(entity: EntityLivingBase, pe: PotionEffect, update: Boolean): Change(entity, pe, update)
	}
	
	abstract class Remove(entity: EntityLivingBase, pe: PotionEffect): LivingPotionEvent(entity, pe) {
		
		@Cancelable
		@Deprecated("Unimplemented", level = DeprecationLevel.ERROR)
		class Pre(entity: EntityLivingBase, pe: PotionEffect): Remove(entity, pe)
		
		class Post(entity: EntityLivingBase, pe: PotionEffect): Remove(entity, pe)
	}
}
